File: peerprint/server/registry.py
import sqlite3
import uuid
from pathlib import Path
from dataclasses import dataclass
import peerprint.server.pkg.proto.peers_pb2 as ppb
from .proc import ProcessOptsBase, ServerProcess
import logging
import tempfile
import time

@dataclass
class RegistryOpts(ProcessOptsBase):
    addr: str = None
    rendezvous: str = None
    local: bool = None
    db: str = None
    connectTimeout: str = None
    syncPeriod: str = None
    lockfile: str = None

class RegistryReader:
    def __init__(self, dbPath, timeout):
        start = time.time()
        while True:
            if time.time() > start+timeout:
                raise Exception("timeout waiting for path " + dbPath)
            if Path(dbPath).exists():
                break
            time.sleep(0.1)

        dbPath = 'file:' + dbPath
        print("Connecting to ", dbPath)
        self.con = sqlite3.connect(dbPath)
        self.con.execute('pragma journal_mode=wal')

        while True:
            if time.time() > start+timeout:
                raise Exception("timeout waiting for DB tables")
            try: 
                self.con.execute("SELECT * FROM schemaversion LIMIT 1")
            except sqlite3.OperationalError:
                time.sleep(0.1)
                continue
            break

    def _toNetwork(self, vals):
        if vals is None:
            return None
        r = ppb.NetworkConfig()
        r.uuid = vals[0]
        r.name = vals[1]
        r.description = vals[2]
        for v in vals[3].split("\n"):
            r.tags.append(v)
        for v in vals[4].split("\n"):
            r.links.append(v)
        r.location = vals[5]
        r.rendezvous = vals[6]
        r.creator = vals[7]
        r.created = vals[8]
        sig = vals[9]
        # Skip stats UUID
        s = ppb.NetworkStats()
        s.population = vals[11]
        s.completions_last7days = vals[12]
        s.records = vals[13]
        s.idle_records = vals[14]
        s.avg_completion_time = vals[15]
        return ppb.Network(config=r, stats=s, signature=sig)

    def _all(self, *args):
        cur = self.con.cursor()
        try:
            return cur.execute(*args).fetchall()
        finally:
            cur.close()

    def getLobby(self):
        return [self._toNetwork(r) for r in self._all("SELECT N.*, S.* FROM lobby N LEFT JOIN stats S ON N.uuid=S.uuid")]

    def register(self, conf, sig):
        with self.con:
            return self.con.execute("INSERT OR REPLACE INTO registry (uuid, name, description, tags, links, location, rendezvous, creator, created, signature) VALUES (?,?,?,?,?,?,?,?,?,?);",  (
                    conf.uuid,
                    conf.name,
                    conf.description,
                    "\n".join(conf.tags),
                    "\n".join(conf.links),
                    conf.location, 
                    conf.rendezvous,
                    conf.creator,
                    conf.created,
                    sig
                ))


class P2PRegistry():
    def __init__(self, opts, binary_path, logger):
        self._logger = logger
        self._opts = opts
        self._binary_path = binary_path

        self._tmpdir = tempfile.TemporaryDirectory()
        if self._opts.db is None:
            self._opts.db = f"{self._tmpdir.name}/state.db"

    def connect(self, timeout):
        self._proc = ServerProcess(self._opts, self._binary_path, self._logger.getChild("proc"))
        self.db = RegistryReader(self._opts.db, timeout)

if __name__ == "__main__":
    logging.basicConfig(level=logging.DEBUG)
    # No lockfile to allow for concurrent test
    reg = P2PRegistry(RegistryOpts(lockfile=""), "./registry", logging.getLogger())
    reg.connect(5.0)
    print("Registering test config")
    cur = reg.db.register(ppb.NetworkConfig(uuid=str(uuid.uuid4()), name="bar", description="baz"), "")
    print("Insert:", cur.rowcount, "rows")
    while True:
        time.sleep(5.0)
        print([n.config.uuid for n in reg.db.getLobby()])

File: peerprint/pkg/storage/registry.go
// Functions for managing a network registry
package storage

import (
  "fmt"
  "strings"
  "log"
  pb "github.com/smartin015/peerprint/p2pgit/pkg/proto"
  "database/sql"
  "context"
  _ "github.com/mattn/go-sqlite3"
  _ "embed"
)

const (
  RegistryTable = "registry"
  LobbyTable = "lobby"
)

//go:embed registry_schema.sql
var registrySchema string

type registry struct {
  path string
  db *sql.DB
}

func NewRegistry(path string) (*registry, error) {
  if path != ":memory:" {
    path = "file:" + path + "?_journal_mode=WAL"
  }
  log.Println("NewRegistry: " + path)
  db, err := sql.Open("sqlite3", path)
  if err != nil {
    return nil, fmt.Errorf("failed to open db at %s: %w", path, err)
  }
  db.SetMaxOpenConns(1) // TODO verify needed for non-inmemory
  s := &registry{
    path: path,
    db: db,
  }
  if err := s.createRegistryTables(); err != nil {
    return nil, fmt.Errorf("failed to create tables: %w", err)
  }
  return s, nil
}


func (s *registry) createRegistryTables() error {
  ver := ""
  if err := s.db.QueryRow("SELECT * FROM schemaversion LIMIT 1;").Scan(&ver); err != nil && err != sql.ErrNoRows && err.Error() != "no such table: schemaversion" {
    return fmt.Errorf("check version: %w", err)
  }

  if ver == "" {
    if _, err := s.db.Exec(string(registrySchema)); err != nil {
      return fmt.Errorf("create tables: %w", err)
    }
    if _, err := s.db.Exec(`INSERT INTO schemaversion (version) VALUES ("0.0.1");`); err != nil {
      return fmt.Errorf("write schema version: %w", err)
    }
  } else {
    fmt.Errorf("Schema version %s", ver);
  }
  return nil
}

func (s *registry) Close() {
  s.db.Close()
}


func (s *registry) DeleteConfig(uuid string, tbl string) error {
  _, err := s.db.Exec(`DELETE FROM ` + tbl + ` WHERE uuid=?`, uuid)
  if err != nil {
    return fmt.Errorf("DeleteConfig: %w", err)
  }
  return nil
}

func (s *registry) UpsertConfig(n *pb.NetworkConfig, sig []byte, tbl string) error {
  _, err := s.db.Exec(`
    INSERT OR REPLACE INTO ` + tbl + ` (uuid, name, description, tags, links, location, rendezvous, creator, created, signature) VALUES (?,?,?,?,?,?,?,?,?,?)`,
    n.Uuid, n.Name, n.Description,
    strings.Join(n.Tags, "\n"),
    strings.Join(n.Links, "\n"),
    n.Location, n.Rendezvous,
    n.Creator, n.Created,
    sig,
  )
  if err != nil {
    return fmt.Errorf("Upsert to registry: %w", err)
  }
  return nil
}

func (s *registry) UpsertStats(uuid string, stats *pb.NetworkStats) error {
  _, err := s.db.Exec(`
    INSERT OR REPLACE INTO stats (uuid, population, completions_last7days, records, idle_records, avg_completion_time) VALUES (?,?,?,?,?,?)`,
    uuid,
    stats.Population,
    stats.CompletionsLast7Days,
    stats.Records,
    stats.IdleRecords,
    stats.AvgCompletionTime,
  )
  if err != nil {
    return fmt.Errorf("Set stats: %w", err)
  }
  return nil
}

func (s *registry) GetLobby(ctx context.Context, cur chan<- *pb.NetworkConfig) error {
  defer close(cur)
  rows, err := s.db.Query(`SELECT * FROM "lobby"`)
  if err != nil {
    return fmt.Errorf("GetLobby SELECT: %w", err)
  }
  defer rows.Close()

  for rows.Next() {
    select {
    case <-ctx.Done():
      return fmt.Errorf("Context canceled")
    default:
    }
    c := &pb.NetworkConfig{}
    tagstr := ""
    linkstr := ""
    sig := ""
    if err := rows.Scan(
      &c.Uuid,
      &c.Name,
      &c.Description,
      &tagstr,
      &linkstr,
      &c.Location,
      &c.Rendezvous,
      &c.Creator,
      &c.Created,
      &sig); err != nil {
      return fmt.Errorf("GetNetworks scan: %w", err)
    }
    c.Tags = strings.Split(tagstr, "\n")
    c.Links = strings.Split(linkstr, "\n")
    cur<- c
  }
  return nil
}


func (s *registry) GetNetworks(ctx context.Context, cur chan<- *pb.Network, closeChan bool) error {
  if closeChan {
    defer close(cur)
  }
  rows, err := s.db.Query(`SELECT R.*, S.* FROM "registry" R LEFT JOIN "stats" S ON S.uuid=R.uuid`)
  if err != nil {
    return fmt.Errorf("GetNetwork SELECT: %w", err)
  }
  defer rows.Close()

  for rows.Next() {
    select {
    case <-ctx.Done():
      return fmt.Errorf("Context canceled")
    default:
    }
    n := &pb.Network{
      Config: &pb.NetworkConfig{},
      Stats: &pb.NetworkStats{},
    }
    tagstr := ""
    linkstr := ""
    var statuuid sql.NullString
    var statpop, statcpl7, statrec, statidle, statavgcpl sql.NullInt64

    if err := rows.Scan(
      &n.Config.Uuid,
      &n.Config.Name,
      &n.Config.Description,
      &tagstr,
      &linkstr,
      &n.Config.Location,
      &n.Config.Rendezvous,
      &n.Config.Creator,
      &n.Config.Created,
      &n.Signature,
      &statuuid,
      &statpop,
      &statcpl7,
      &statrec,
      &statidle,
      &statavgcpl); err != nil {
      return fmt.Errorf("GetNetworks scan: %w", err)
    }
    n.Config.Tags = strings.Split(tagstr, "\n")
    n.Config.Links = strings.Split(linkstr, "\n")
    n.Stats.Population = statpop.Int64
    n.Stats.CompletionsLast7Days = statcpl7.Int64
    n.Stats.Records = statrec.Int64
    n.Stats.IdleRecords = statidle.Int64
    n.Stats.AvgCompletionTime = statavgcpl.Int64
    cur<- n
  }
  return nil
}

func (s *registry) SignConfig(uuid string, sig []byte) error {
  _, err := s.db.Exec(`UPDATE stats SET signature=? WHERE uuid=?`, sig, uuid)
  if err != nil {
    return fmt.Errorf("SignConfig: %w", err)
  }
  return nil
}
